cirq/sim/act_on_density_matrix_args.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Objects and methods for acting efficiently on a density matrix."""
+
+from typing import Any, Iterable, Dict, List, Sequence, Tuple
+
+import numpy as np
+
+from cirq import protocols, ops
+from cirq.protocols.decompose_protocol import _try_decompose_into_operations_and_qubits
+
+
+class ActOnDensityMatrixArgs:
+    """State and context for an operation acting on a density matrix.
+
+    There are three common ways to act on this object:
+
+    1. Directly edit the `target_tensor` property, which is storing the density
+        matrix of the quantum system as a numpy array with one axis per qudit.
+    2. Overwrite the `available_buffer` property with the new state vector, and
+        then pass `available_buffer` into `swap_target_tensor_for`.
+    3. Call `record_measurement_result(key, val)` to log a measurement result.
+    """
+
+    def __init__(
+        self,
```

Comment:
A good chunk of this was copied straight from state vector args. Probably worth making a base class. That would allow some deduplication in e.g. MeasurementGate too.

Reply:
Initial thought is that would be better as a separate PR.

Reply:
Nvm, I went ahead and did it here, it wasn't much trouble.